fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordContext.java

```diff
@@ -221,28 +227,115 @@ public long getTransactionCreateTime() {
+         * @return {@code true} if the check is complete
+         */
+        default boolean isReady() {
+            return false;
+        }
+
+        /**
+         * Complete the check.
+         *
+         * This is always called once before {@link #commit} completes. If {@link #isReady} returns {@code true} earlier,
+         * it can be called while processing the transaction.
+         * @return a future that will be complete (exceptionally if the check fails) when the check has been performed
+         */
+        @Nonnull
+        CompletableFuture<Void> checkAsync();
+    }
+
+    /**
+     * A synchronous {@link CommitCheckAsync}.
+     *
+     * At some point, this class will be deprecated.
+     * Please implement {@link CommitCheckAsync} directly or call {@link #addCommitCheck(CompletableFuture)} instead.
      */
-    public interface CommitCheck {
-        boolean isReady();
+    public interface CommitCheck extends CommitCheckAsync {
+        @Override
+        @Nonnull
+        default CompletableFuture<Void> checkAsync() {
+            check();
+            return AsyncUtil.DONE;
+        }
 
+        /**
+         * Complete the check.
+         *
+         * This is always called once before {@link #commit} completes. If {@link #isReady} returns {@code true} earlier,
+         * it can be called while processing the transaction.
+         *
+         * <p>
+         * This method should not block or {@link #commitAsync} will block. It is therefore much
+         * better to always implement {@link CommitCheckAsync} or call {@link #addCommitCheck(CompletableFuture)} instead.
+         */
         void check();
     }
 
-    public synchronized void addCommitCheck(@Nonnull CommitCheck check) {
+    /**
+     * Add a {@link CommitCheckAsync} to be performed before {@link #commit} finishes.
+     *
+     * This method is suitable for checks that cannot be started until just before commit.
+     * For checks that can be started before {@code addCommitCheck} time, {@link #addCommitCheck(CompletableFuture)}
+     * may be more convenient.
+     * <p>
+     * It is possible for this method to throw an exception caused by an earlier unsuccessful check that has become ready in the meantime.
+     * @param check the check to be performed
+     */
+    public synchronized void addCommitCheck(@Nonnull CommitCheckAsync check) {
         while (!commitChecks.isEmpty()) {
             if (commitChecks.peek().isReady()) {
-                commitChecks.remove().check();
+                asyncToSync(FDBStoreTimer.Waits.WAIT_ERROR_CHECK, commitChecks.remove().checkAsync());
             } else {
                 break;
             }
         }
         commitChecks.add(check);
     }
 
-    public synchronized void runCommitChecks() {
-        while (!commitChecks.isEmpty()) {
-            commitChecks.remove().check();
+    /**
+     * Add a check to be completed before {@link #commit} finishes.
```

Comment:
Does the link work, or does it need to be `{@link #commit()}`?

Reply:
Yes, it works in both IntelliJ and JavaDoc. In the latter, the result is
```
... before <a href="..."><code>commit()</code></a> finishes ...
```

Reply:
Okay, that's fine.